SDWebImageFLPlugin/Classes/FLAnimatedImageBridge/SDWebImageFLCoder.m

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of the SDWebImage package.
+ * (c) Olivier Poitrey <rs@dailymotion.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+#import "SDWebImageFLCoder.h"
+#import "FLAnimatedImageView+WebCache.h"
+
+@implementation SDWebImageFLCoder
+
++ (SDWebImageFLCoder *)sharedCoder {
+    static SDWebImageFLCoder *coder;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        coder = [[SDWebImageFLCoder alloc] init];
+    });
+    return coder;
+}
+
+- (BOOL)canDecodeFromData:(NSData *)data {
+    return ([NSData sd_imageFormatForImageData:data] == SDImageFormatGIF);
+}
+
+- (UIImage *)decodedImageWithData:(NSData *)data options:(SDImageCoderOptions *)options {
+    SDWebImageContext *context = options[SDImageCoderWebImageContext];
+    NSString *operationKey = context[SDWebImageContextSetImageOperationKey];
+    Class imageViewClass = NSClassFromString(operationKey);
+    
+    // Check if image request come from `FLAnimatedImageView`
+    if (imageViewClass && [imageViewClass isSubclassOfClass:FLAnimatedImageView.class]) {
+        // Parse args
+        BOOL predrawingEnabled = YES;
+        if (context[SDWebImageContextPredrawingEnabled]) {
+            predrawingEnabled = [context[SDWebImageContextPredrawingEnabled] boolValue];
+        }
+        NSUInteger optimalFrameCacheSize = 0;
+        if (context[SDWebImageContextOptimalFrameCacheSize]) {
+            optimalFrameCacheSize = [context[SDWebImageContextOptimalFrameCacheSize] unsignedIntegerValue];
+        }
+        // Create FLAnimatedImage
+        FLAnimatedImage *animatedImage = [[FLAnimatedImage alloc] initWithAnimatedGIFData:data optimalFrameCacheSize:optimalFrameCacheSize predrawingEnabled:predrawingEnabled];
+        if (!animatedImage) {
+            return nil;
+        }
+        
+        return [UIImage sd_imageWithFLAnimatedImage:animatedImage];
+    } else {
+        UIImage *image;
+        NSArray<id<SDImageCoder>> *coders = [SDImageCodersManager sharedManager].coders;
```

Comment:
@dreampiggy TBO, I don't much like handle like this, it's a little bit non-decouple very well, my opinion is if `SDWebImageFLCoder` or any other coder can't create the image, just return `nil`, `SDImageCodersManager` can check if it is `nil`, it would enumerate next coder to check wether can decode. I can make a `PR` if you think it deserve. 😂 
What's your opinion? 🤔

Reply:
The image coder protocol, it's not designed for `SDImageCodersManager`...Actually.

The `decodedImageWithData:options:` return nil, does not means : `This coder can not support this compressed image foramt`.

For example, during image decoding, you are trying to parse the bianry frame hex for target image format, but it failed.
For example, during allocation of large memory to outpout bitmap buffer, it failed.
For example, during animation frame looping, it failed.

These cases, should not call another coder. Because it's a internal error happend for that coder, but not image foramt.

However, the `canDecodeImageWithData:` it's about image format check, not the decoding issue. Actually it should be named with `canDecodedImageFormat:`, but using a `NSData *` make it more extensible.

Reply:
@dreampiggy I suddenly realized that maybe `Apple` do the same as my thought. Apple has a manager to figure out what the image format the data is, if it is `JPEG`, it would call `JPEGPlugin` to decode, if it is `PNG`, it would call `PNGPlugin`. Other than enumerate all plugins to ask them wether they can decode the data. Just like broadcast vs peer to peer.

Reply:
@zhongwuzw No. ImageIO is just a Core Foundation wrapper for the C++ framework IIOReadPlugin && IIOWritePlugin. It actually use the same design from SDWebImage's coder (Maybe because that I "copy" the design from their implementation :))

IIOReadPlugin use a array (`std::vector`) to store each of plugins (Which subclass of IIOReadPlugin abstract class). And for each plugin, call their handle method instead. When you call `CGImageSourceCreateWithXXX`, it will try to bind the Core Foundation `CGImageSourceRef` class to a actual C++ `IIOReadPlugin` internally. They just do the for-in loop as we can (Until you specify a UTType during `CGImageSourceCreateWithXXX` method)


The good point it's that they can hard-code all the plugins into single files, but we can't. We open the posibility to let other third-party to provide a coder plugin. So this is why we can't do the same thing to hold a central check of UTType.

If you're interested in the detailed implementation, You can use hopper to deassemble ImageIO.framework. Here is just some snippet from my previous experience.

Reply:
For example, they do for-loop to check 16-bytes only of File Signature (See above) to determine which plugin should be binded for current `CGImageSource` during `IIO_ReaderHandler::readerForBytes()`. There are also some other naive check like `IIO_ReaderHandler::readerForUTType()` and `IIO_ReaderHandler::readerForPathExtension()`

Reply:
@dreampiggy Maybe you are more familiar than me at `ImageIO`, I just see it start last hours before 😂 . Am I missed something? I profile using Instruments first, the result is below, ` IIOImageSource::createImageAtIndex` would use `CGImageSource` to create `CGImageRef`, then I trace the steps, I didn't find any `for loop` to find the plugin. I can step into it more detail carefully, may I miss anything?

I use `CGImageSourceCreateWithData` and `CGImageSourceCreateImageAtIndex` to create image.

```
303.00 ms   22.2%	303.00 ms	 	 aj_icol_mcurow_cmyk
303.00 ms   22.2%	0 s	 	  aj_bufferproc_crop
303.00 ms   22.2%	0 s	 	   aj_decode_all
303.00 ms   22.2%	0 s	 	    aj_decode_all_mt
303.00 ms   22.2%	0 s	 	     applejpeg_decode_image_all
303.00 ms   22.2%	0 s	 	      AppleJPEGReadPlugin::copyImageBlockSet(InfoRec*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*)
303.00 ms   22.2%	0 s	 	       AppleJPEGReadPlugin::CopyImageBlockSetProc(void*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*)
303.00 ms   22.2%	0 s	 	        IIOImageProviderInfo::copyImageBlockSetWithOptions(CGImageProvider*, CGRect, CGSize, __CFDictionary const*)
303.00 ms   22.2%	0 s	 	         IIOImageProviderInfo::CopyImageBlockSetWithOptions(void*, CGImageProvider*, CGRect, CGSize, __CFDictionary const*)
303.00 ms   22.2%	0 s	 	          CGImageProviderCopyImageBlockSetWithOptions
159.00 ms   11.6%	0 s	 	           invocation function for block in IIOReadPlugin::cacheImmediately(__CFDictionary const*, CGImage*)
159.00 ms   11.6%	0 s	 	            _dispatch_client_callout
159.00 ms   11.6%	0 s	 	             _dispatch_queue_barrier_sync_invoke_and_complete
159.00 ms   11.6%	0 s	 	              IIOReadPlugin::cacheImmediately(__CFDictionary const*, CGImage*)
159.00 ms   11.6%	0 s	 	               IIOImageSource::createImageAtIndex(unsigned long, IIODictionary*)
159.00 ms   11.6%	0 s	 	                CGImageSourceCreateImageAtIndex
```

Reply:
@zhongwuzw ...Paste many deassemble code here seems contains no use. You can just use Hopper, search ImageIO.framework with that symbols. 

Each plugins, in `IIO_ReaderHandler::buildPluginList()`, call their create method, like `CreateReader_HEIC`. In the `CreateReader_HEIC`, it create a new plugin which is subclass of `IIOGeneric_Reader`. Then register the plugin with their desired UTType (For HEIC, it's the `_kCGImageTypeIdentifierHEIC`)

During CGImageSourceGetCount / CGImageSourceCreateImageAtIndex and all the process method, it will lazy check whether a plugin has been binded to CGImageSource using ` IIOImageSource::bindToReader()`. It will called only once (For non-progressive decoding) and use `IIO_ReaderHandler::readerForBytes()` to check File Signature, convert to UTType, and finally find the plugin that match the UTType.

Anyway, it's just a personal experience, and deassemble code may not been considered  a legal way of understanding the internal behavior for their framework. If you're interested in, just do yourself. :)

```cpp
int __ZN17IIO_ReaderHandler15readerForUTTypeEPK10__CFString(void * arg0) {
    r15 = rsi;
    r12 = arg0;
    rbx = *(r12 + 0x10);
    if (rbx == *(r12 + 0x18)) goto loc_10cc0e;

loc_10cbe5:
    r14 = 0x0;
    goto loc_10cbe8;

loc_10cbe8:
    if (CFStringCompare(**(*rbx + 0x8), r15, 0x0) == 0x0) goto loc_10cc13;

loc_10cc01:
    rbx = rbx + 0x8;
    if (rbx != *(r12 + 0x18)) goto loc_10cbe8;

loc_10cc16:
    rax = r14;
    return rax;

loc_10cc13:
    r14 = *rbx;
    goto loc_10cc16;

loc_10cc0e:
    r14 = 0x0;
    goto loc_10cc16;
}

```